### PR TITLE
add debug build target

### PR DESCRIPTION
**Pull request by u8-Salem**
Adds a debug build target

## Description
`make debug` will build separate artifacts in `build/(modern-)debug` and enable `-Og`  and `-g`.
for agbcc it will enable `-O1` and `-g`
Also adds the option to run any build with `NOOPT=1` remove any optimization flags

sets all debug menu options in `config/debug.h` to false by default. the debug target turns all of these options on by default. 
To achieve this, any preproc directive regarding these menus has been removed and integrated in the regular control flow. 

Also fixes an error that reveals when building with debug (Werror=maybe-uninitialized)

## **Discord contact info**
u8.salem

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -44,6 +44,7 @@ MODERN       ?= 1
 TEST         ?= 0
 ANALYZE      ?= 0
 UNUSED_ERROR ?= 0
+DEBUG		 ?= 0
```
Comment (Bassoonian) [inline Makefile:47]: spaces instead of tabs

File: include/config.h [hunk 2]
```diff
@@ -8,6 +8,10 @@
 // Ruby's actual debug build does not use the AGBPrint features.
 #define NDEBUG
 
+#ifdef DEBUG_TARGET
+#undef NDEBUG
+#endif
+
 // To enable printf debugging, comment out "#define NDEBUG". This allows
```
Comment (Bassoonian) [inline include/config.h:15]: I wonder if the comment needs to be changed to accomodate the new debug build target?

File: src/battle_controller_player.c [hunk 3]
```diff
@@ -413,13 +413,11 @@ static void HandleInputChooseAction(u32 battler)
     {
         SwapHpBarsWithHpText();
     }
-#if DEBUG_BATTLE_MENU == TRUE
-    else if (JOY_NEW(SELECT_BUTTON))
+    else if (JOY_NEW(SELECT_BUTTON) && DEBUG_BATTLE_MENU)
```
Comment (Bassoonian) [inline src/battle_controller_player.c:416]:
```suggestion
    else if (DEBUG_BATTLE_MENU && JOY_NEW(SELECT_BUTTON))
```

File: src/battle_main.c [hunk 4]
```diff
@@ -515,9 +515,7 @@ static void CB2_InitBattleInternal(void)
     gBattle_BG3_X = 0;
     gBattle_BG3_Y = 0;
 
-#if DEBUG_OVERWORLD_MENU == TRUE
-    if (!gIsDebugBattle)
-#endif
+    if ((DEBUG_OVERWORLD_MENU && !gIsDebugBattle) || !DEBUG_OVERWORLD_MENU)
```
Comment (Bassoonian) [inline src/battle_main.c:518]:
```suggestion
    if (!DEBUG_OVERWORLD_MENU || (DEBUG_OVERWORLD_MENU && !gIsDebugBattle))
```

File: src/battle_main.c [hunk 5]
```diff
@@ -550,9 +548,7 @@ static void CB2_InitBattleInternal(void)
     else
         SetMainCallback2(CB2_HandleStartBattle);
 
-#if DEBUG_OVERWORLD_MENU == TRUE
-    if (!gIsDebugBattle)
-#endif
+    if ((DEBUG_OVERWORLD_MENU && !gIsDebugBattle) || !DEBUG_OVERWORLD_MENU)
```
Comment (Bassoonian) [inline src/battle_main.c:551]:
```suggestion
    if (!DEBUG_OVERWORLD_MENU || (DEBUG_OVERWORLD_MENU && !gIsDebugBattle))
```

File: src/field_control_avatar.c [hunk 6]
```diff
@@ -200,15 +201,16 @@ int ProcessPlayerFieldInput(struct FieldInput *input)
...
 
-#if DEBUG_OVERWORLD_MENU == TRUE && DEBUG_OVERWORLD_IN_MENU == FALSE
-    if (input->input_field_1_2)
+    if(DEBUG_OVERWORLD_MENU && !DEBUG_OVERWORLD_IN_MENU)
     {
-        PlaySE(SE_WIN_OPEN);
-        FreezeObjectEvents();
-        Debug_ShowMainMenu();
-        return TRUE;
+        if (input->input_field_1_2)
```
Comment (Bassoonian) [inline src/field_control_avatar.c:206]: I would just add this to the preceding if statement. No need to stack them like this

File: src/pokemon_summary_screen.c [hunk 7]
```diff
@@ -1640,15 +1640,13 @@ static void Task_HandleInput(u8 taskId)
             PlaySE(SE_SELECT);
             BeginCloseSummaryScreen(taskId);
         }
-    #if DEBUG_POKEMON_MENU == TRUE
-        else if (JOY_NEW(SELECT_BUTTON) && !gMain.inBattle)
+        else if (JOY_NEW(SELECT_BUTTON) && !gMain.inBattle && DEBUG_POKEMON_MENU)
```
Comment (Bassoonian) [inline src/pokemon_summary_screen.c:1643]:
```suggestion
        else if (DEBUG_POKEMON_MENU && JOY_NEW(SELECT_BUTTON) && !gMain.inBattle)
```

File: include/config.h (hunk 2 above)
Comment (u8-Salem) [reply include/config.h:15]: I dont know if `#undef``'ing NDEBUG when its not defined is UB? we could add smth like `#ifdef NDEBUG` to catch the case that it is already commented out.

Comment (mrgriffin) [reply include/config.h:15]: It's safe to `#undef` things that aren't defined: [cppreference](https://en.cppreference.com/w/cpp/preprocessor/replace#.23undef_directive).

Comment (u8-Salem) [reply include/config.h:15]: then I am not sure what should be changed about the comment. A note that the same can be achieved using debug target perhaps ?

Comment (u8-Salem) [reply include/config.h:15]: @Bassoonian did you want a change here still ?

Comment (Bassoonian) [reply include/config.h:15]: Nah it's fine, no worries! I don't feel equipped enough to approve and merge this PR, though, so I'll leave that to the Makefile gods :)

**Pull request closed without merging**